### ISSUE TITLE
Feature: custom tree view of current project

### ISSUE THREAD

**Issue by rmunn** [ionide/ionide-vscode-fsharp#248]
Once https://github.com/Microsoft/vscode/pull/14048 is accepted and merged, VS Code will allow extensions to create a custom tree explorer. I'd like to see a visual representation of the .fsproj file -- especially the compilation order of the files in the project. I'd like to be able to click on a file and use `Alt+Up` or `Alt+Down` to rearrange it in project order, with visual feedback. (Currently the "F#: Move File Up" and "F#: Move File Down" commands give no visual feedback -- so I'm never sure if they're worked or not.) I'd also like to be able to right-click on a project and choose "Add New File..." or "Add Existing File..." from a popup menu. It might be possible to add more features later (like right-clicking to add project references, which would call Paket with appropriate parameters). But just having a visual representation of the compilation order of my files, and being able to rearrange them without having to look at the ugly XML syntax of .fsproj files, would be all I really want right now.

**Comment by Krzysztof-Cieslak** [comment]
Yes, we know about this PR and monitor it. Using it for better project explorer is definitely something we want to do.

**Comment by Krzysztof-Cieslak** [comment]
Prototype on `workspaceTree` branch.

**Comment by enricosada** [comment]
@Krzysztof-Cieslak you need some info (graph of `<Compile`, `None`, `Content` items, anything) to do that with new fsproj? 

i can do that easy for readonly (no xml parsing, real msbuild evaluation, like current integration).

`NewFsprj -> MsbuildItem list` :

```fsharp
type MsbuildItem =
   | Folder of (string: name) * (MsbuildItem list)
   | None of (name: string) * (path: string)
   | Compiler of (name: string) * (path: string)
   | Content of (name: string) * (path: string)
```

Update is not something i can do now, but new fsproj is easy to manually edit

**Comment by Krzysztof-Cieslak** [comment]
Main problem is that usable API on Code side is not released yet *shrug*

**Comment by Kurren123** [comment]
@Krzysztof-Cieslak are there any issues I can monitor to track whether a usable API is coming? IMO this is the biggest thing that Visual Studio has over VS code (for F#) at the moment. If we can make this happen then there will be almost no reason to use Visual Studio

**Comment by Krzysztof-Cieslak** [comment]
![capture](https://cloud.githubusercontent.com/assets/5427083/26494152/c461740e-421d-11e7-9d8e-6efd2fed86d9.PNG)

Prototype using new tree API from https://github.com/Microsoft/vscode/issues/26948

**Comment by enricosada** [comment]
Awesome @Krzysztof-Cieslak !!

**Comment by Kurren123** [comment]
It's... beautiful

**Comment by Krzysztof-Cieslak** [comment]
![ionide-project](https://cloud.githubusercontent.com/assets/5427083/26513058/c6830dbe-4269-11e7-8a70-5d4f506157da.gif)

**Comment by forki** [comment]
:kiss:

**Comment by 7sharp9** [comment]
How do you select what project you are on?

**Comment by Krzysztof-Cieslak** [comment]
Right now it displays all projects that are in current workspace (currently opened folder in Code)

**Comment by Krzysztof-Cieslak** [comment]
Initial version implemented in `2.26.1`
